src/dwi/tractography/seeding/basic.h

```diff
@@ -160,6 +161,119 @@ namespace MR
 
 
 
+        class Coordinate_parser
+        { MEMALIGN(Coordinate_parser)
+
+          public:
+
+            Eigen::MatrixXf coords;
+            size_t nr;
+            size_t nc;         
+
+            Coordinate_parser (const std::string& cds_path) {
+              
+              coords = load_matrix<float> (cds_path);
+              nr = coords.rows();
+              nc = coords.cols();
+
+            }  
+
+        };
+
+
+        
+        class Coordinates_fixed : public Base, public Coordinate_parser
+        { MEMALIGN(Coordinates_fixed)
+
+          public:
+
+            Coordinates_fixed (const std::string& in, const size_t n_streamlines) :
+              Base (in, "coordinate seeding fixed", MAX_TRACKING_SEED_ATTEMPTS_FIXED), 
+              Coordinate_parser(in) {
+
+                if (nc != 3)
+                  throw Exception ("Number of columns in \"" + in + "\" must equal 3!");   
+                if (n_streamlines == 0)
+                  throw Exception ("Number of streamlines to be generated per coordinate must be a positive integer!");             
+                current_coord = 0;
+                num_at_coord = 0;
+                expired = false;
+                nsl = n_streamlines;
+                count = nr * nsl;
+
+              }
+
+            virtual bool get_seed (Eigen::Vector3f& p) const override;
+
+          private:
+
+            mutable size_t current_coord;
+            mutable size_t num_at_coord;
+            mutable bool expired;   
+            size_t nsl;            
+
+        };
+
+
+
+        class Coordinates_global : public Base, public Coordinate_parser
+        { MEMALIGN(Coordinates_global)
+
+          public:
+
+            Coordinates_global (const std::string& in) :
+              Base (in, "coordinate seeding global", MAX_TRACKING_SEED_ATTEMPTS_RANDOM), 
+              Coordinate_parser(in) {
+
+                if (nc < 3 && nc > 4)
+
+                  throw Exception ("Number of columns in \"" + in + "\" must equal 3 or 4!"); 
+
+                if (nc == 4) {
+
+                  if (coords.col(3).minCoeff() < 0)
+                    throw Exception ("The seeding weights must be non-negative!");
+
+                  sorted_weights = coords.col(3);
+                  std::sort(sorted_weights.data(), sorted_weights.data() + nr);
+
+                  visited = Eigen::MatrixXf::Zero(nr, 1);
```

Comment:
Constructor doesn't match variable type definition.

Reply:
Also, appears to only be used in the constructor, not elsewhere; therefore it can be defined as a local variable rather than a class member.

Reply:
Also `vector<bool>` would be better.

Reply:
Also `bool` would be better than `float`.